### ISSUE TITLE
Code is not cheked after some function calls: binder/inference bug

### ISSUE THREAD

**Issue by ilevkivskyi** [python/mypy#3994]
Any code after an ``UninhabitedType`` is marked as unreachable in binder. This leads to dangerous behaviour together with how type inference currently works "in all uncertain situations infer uninhabited" (see also https://github.com/python/mypy/issues/3283 for discussion). For example no error is reported in this code:
```python
from typing import TypeVar
T = TypeVar('T')

def f() -> T:
    ...

class C:
    f()
    1 + 'this is not checked'

def fun() -> None:
    f()
    1 + 'can do everything here'

if True:
    f()
    1 + 'whatever'
```
There are other function signatures (all involve type variables) that lead to this behaviour: type checking stops after a call to such function.

**Comment by elazarg** [comment]
I think that in a dynamic settings `()->T` should not be considered as returning `Uninhabited`. Similarly to how `T->T` is not necessarily the identity function.
